core/src/main/java/org/elasticsearch/index/query/HasChildQueryBuilder.java

```diff
@@ -120,4 +241,236 @@ protected void doXContent(XContentBuilder builder, Params params) throws IOExcep
+                if (innerHits != null) {
+                    ParsedQuery parsedQuery = new ParsedQuery(innerQuery, context.copyNamedQueries());
+                    InnerHitsContext.ParentChildInnerHits parentChildInnerHits = new InnerHitsContext.ParentChildInnerHits(innerHits.getSubSearchContext(), parsedQuery, null, context.mapperService(), childDocMapper);
+                    String name = innerHits.getName() != null ? innerHits.getName() : type;
+                    context.addInnerHits(name, parentChildInnerHits);
+                }
+            }
+        }
+
+        String parentType = parentFieldMapper.type();
+        DocumentMapper parentDocMapper = context.mapperService().documentMapper(parentType);
+        if (parentDocMapper == null) {
+            throw new QueryShardException(context, "[" + NAME + "] Type [" + type + "] points to a non existent parent type ["
+                    + parentType + "]");
+        }
+
+        if (maxChildren > 0 && maxChildren < minChildren) {
+            throw new QueryShardException(context, "[" + NAME + "] 'max_children' is less than 'min_children'");
+        }
+
+        BitSetProducer nonNestedDocsFilter = null;
+        if (parentDocMapper.hasNestedObjects()) {
+            nonNestedDocsFilter = context.bitsetFilter(Queries.newNonNestedFilter());
+        }
+
+        // wrap the query with type query
+        innerQuery = Queries.filtered(innerQuery, childDocMapper.typeFilter());
+
+        final Query query;
+        final ParentChildIndexFieldData parentChildIndexFieldData = context.getForField(parentFieldMapper.fieldType());
+        if (context.indexVersionCreated().onOrAfter(Version.V_2_0_0_beta1)) {
+            int maxChildren = maxChildren();
+            // 0 in pre 2.x p/c impl means unbounded
+            if (maxChildren == 0) {
+                maxChildren = Integer.MAX_VALUE;
+            }
+            query = new LateParsingQuery(parentDocMapper.typeFilter(), innerQuery, minChildren(), maxChildren, parentType, scoreTypeToScoreMode(scoreType), parentChildIndexFieldData);
+        } else {
+            // TODO: use the query API
+            Filter parentFilter = new QueryWrapperFilter(parentDocMapper.typeFilter());
+            if (minChildren > 1 || maxChildren > 0 || scoreType != ScoreType.NONE) {
+                query = new ChildrenQuery(parentChildIndexFieldData, parentType, type, parentFilter, innerQuery, scoreType, minChildren,
+                        maxChildren, shortCircuitCutoff, nonNestedDocsFilter);
+            } else {
+                query = new ChildrenConstantScoreQuery(parentChildIndexFieldData, innerQuery, parentType, type, parentFilter,
+                        shortCircuitCutoff, nonNestedDocsFilter);
+            }
+        }
+        if (queryName != null) {
+            context.addNamedQuery(queryName, query);
+        }
```

Comment:
Can be removed here, queryName and boost are all handled already in AbstractQueryBuilder#toQuery().